### PR TITLE
[fix] add data type check to array setter

### PR DESCRIPTION
**Pull request by KexinFeng**
## Description ##

Fix https://github.com/deepjavalibrary/djl/issues/1970. The test code is therein.

## details
A poential place to convert the datatype from the input data type to the target array data type is in
https://github.com/deepjavalibrary/djl/blob/aeb9135686605aab846ada67df800c954b5ebf63/api/src/main/java/ai/djl/ndarray/BaseNDManager.java#L428

Here it does the job of:
> 2\. Multiple dataType maps to the same Buffer type, we have to handle them differently

But it is not ideal to conver the Buffer type inside java
https://stackoverflow.com/questions/38745123/bytebufferasshortbuffer-cannot-be-cast-to-java-nio-floatbuffer
Cannot cast 'java.nio.HeapIntBuffer' to 'java.nio.FloatBuffer'

So here we simply throw an exception when datatypes don't match.

### CODE REVIEW

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java [hunk 1]
```diff
@@ -212,6 +212,16 @@ public NDArray get(NDIndex index) {
     /** {@inheritDoc} */
     @Override
     public void set(Buffer data) {
+        DataType arrayType = getDataType();
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:215]: I don't think we check list this:
1. We should always allow ByteBuffer to set for any DataType
2. Multiple dataType maps to the same Buffer type, we have to handle them differently

Comment (KexinFeng) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:215]: I find that in the implementation of set(Buffer data) in different engines, like MXNet, PyTorch, TensorFlow, the buffer is already converted to ByteBuffer before being fed into the engines. Do you mean we should add this to NDArrayAdapter?

Comment (KexinFeng) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:215]: By check listing those, is it out of the purpose of automatically converting the input datatype into the target array datatype?

**Comment by KexinFeng** [comment]
A poential place to convert he datatype from the input data type to the target array data type is in
https://github.com/deepjavalibrary/djl/blob/aeb9135686605aab846ada67df800c954b5ebf63/api/src/main/java/ai/djl/ndarray/BaseNDManager.java#L428

Here it does the job of:
> 2\. Multiple dataType maps to the same Buffer type, we have to handle them differently

But it is not ideal to conver the Buffer type inside java
https://stackoverflow.com/questions/38745123/bytebufferasshortbuffer-cannot-be-cast-to-java-nio-floatbuffer
Cannot cast 'java.nio.HeapIntBuffer' to 'java.nio.FloatBuffer'

**Review by frankfliu — APPROVED** [review]

File: api/src/main/java/ai/djl/ndarray/BaseNDManager.java [hunk 2]
```diff
@@ -404,9 +404,18 @@ NDManager getAlternativeManager() {
      * @throws IllegalArgumentException if buffer size is invalid
      */
     public static void validateBufferSize(Buffer buffer, DataType dataType, int expected) {
+        boolean isByteBuffer = buffer instanceof ByteBuffer;
+        DataType type = DataType.fromBuffer(buffer);
+        if (!isByteBuffer && type != dataType) {
```
Comment (KexinFeng) [inline api/src/main/java/ai/djl/ndarray/BaseNDManager.java:409]: The following mismatch case will escape the check:
`type != dataType && buffer is ByteBuffer`, but `dataType` is not one of the byte types = {DataType.UINT8, DataType.INT8, DataType.BOOLEAN}

```suggestion

if (arrayType != inputType ) {
          DataType[] types = {DataType.UINT8, DataType.INT8, DataType.BOOLEAN};
          if (!isByteBuffer || Arrays.stream(types).noneMatch(x -> x == dataTypeType)) {
          
```

Comment (frankfliu) [reply api/src/main/java/ai/djl/ndarray/BaseNDManager.java:409]: ByteBuffer should always be allowed. It's the memory representation of all data types. All NDArray has `toByteBuffer()` method. The JNI only accept ByteBuffer, if we only accept matching Buffer, we have to covert ByteBuffer to Buffer and copy the Buffer to ByteBuffer to pass to JNI. It's not efficient for Hybrid engine. We trying to achieve 0 copy between pytorch and onnx engines. It rely on ByteBuffer.

**Pull request merged**